iwyu.cc

```diff
@@ -3358,7 +3362,7 @@ class InstantiatedTemplateVisitor
   // If we're not in the resugar-map, then we weren't canonicalized,
   // so we can just use the input type unchanged.
   const Type* ResugarType(const Type* type) const {
-    type = Desugar(type);
```

Comment:
Before my change to `Desugar` this only did `RemoveSubstTemplateTypeParm`, which I suppose is a form of resugaring. Now that we're going all the way to the canonical type, it's very much the reverse of resugaring.

What would happen if we did something like
```
const Type* canonical_type = GetCanonicalType(type);
if (ContainsKeyValue(resugar_map_, canonical_type, static_cast<const Type*>(nullptr)))
  return type;  // keep the original type if no match
return GetOrDefault(resugar_map_, canonical_type, type);  // keep the original type if no match
```

That seems to me it would return a more sugared type. But on the other hand I suspect it becomes much less deterministic what the result of `ResugarType` is. Thoughts?

Reply:
In fact, the change in this line isn't needed. I've added it only for consistency. I can apply your suggestion or revert it back at all. Not sure.

Reply:
I also appreciate the consistency, so I'm inclined to keep it like it is in this patch.

But are you saying that `ResugarType` is already always called with canonical types? I.e. the following would hold:
```
CHECK_(type == GetCanonicalType(type));
```
? Might be nice to that instead.
```

Reply:
No, I meant that it is called either with canonical type, or with sugared type (from the cache). In the latter case, it just returns back that sugared type.

Reply:
Ah, thanks. As I touched on earlier, I think it makes sense for `ResugarType` to return something deterministic if it's called with a type that is not in the resugar map. And I think using the canonical type is the best way to highlight missing resugaring (as opposed to returning back an accidentally-/half-sugared type). So let's keep this as-is. LGTM.

Reply:
I've recalled that I intend to make `InstantiatedTemplateVisitor` reporting dependent types which are unrelated to original template arguments (see [that comment](https://github.com/include-what-you-use/include-what-you-use/pull/1182#issuecomment-1406681708)). So, returning original type here makes sense. But it could be done along with aforementioned change (and along with writing corresponding test case).